### PR TITLE
Mempool protocol: abstract description [WIP]

### PR DESCRIPTION
**Pull request by cason**
Rendered: https://github.com/cometbft/knowledge-base/blob/cason/mempool-protocol-spec/protocols/mempool/protocol-abstract.md

### CODE REVIEW

File: protocols/mempool/protocol-abstract.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        senders[tx] = append(senders[tx], q)
+
+upon commit(txs):
+    for all tx in txs
+        seen[tx] = true
+        mempool = remove(mempool, tx) // Optional improvement
+
+func send_routine(p):
+    while p in peers
+        tx = mempool.next() // blocks while there is no next
```
Comment (hvanz) [inline protocols/mempool/protocol-abstract.md:205]: While next is blocking, p may be disconnected.

File: protocols/mempool/protocol-abstract.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+> of transaction ordering when sending transactions to peers (either existing or new).
+> However, users of the protocol expect the mempool protocol to address the above
+> two limitations, which can be attested by the only document specifically
+> addressing the mempool in CometBFT documentation:
+> https://github.com/cometbft/cometbft/blob/main/docs/core/mempool.md
+
+
+## Mempool flooding algorithm
+
+The following algorithm is obtained by refining the previous algorithm as follows.
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:126]: "following" is a bit odd here 

In fact, the code below takes into account the mempool itself. So, strictly speaking, this is more than a refinment.

File: protocols/mempool/protocol-abstract.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+The first step is to replace the blocking **send** invocations by non-blocking
+equivalents which _schedule_ the sending of a transaction to a peer.
+The scheduled send calls have to preserve order, in the sense that the sending
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:130]: " scheduled send calls" heavy

File: protocols/mempool/protocol-abstract.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+of two transactions to the same peer must be performed in the same order at
+which they were scheduled.
+
+This first step enables making the sending blocks, in the _for all_ excerpts of
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:134]: "This first step enable"
what do you refer to here?

File: protocols/mempool/protocol-abstract.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+```
+var peers   // other nodes to which we are directly connected
+var seen    // map of transactions already seen
+var mempool // list of transactions
+var senders // map of processes from which transactions were received
+
+upon brodcast(tx):
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        mempool = append(mempool, tx)
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:188]: I prefer OOP here, that is `mempool.append(tx)`. But this might be a matter of taste ^^

Comment (otrack) [reply protocols/mempool/protocol-abstract.md:188]: This is not inline with the codebase: when a client sends a tx to the local node, the tx goes first through the mempool before reaching the reactor. Nonetheless, I believe that this simplification is nice to model things.

File: protocols/mempool/protocol-abstract.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+upon receive <TX, tx> from p:
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        senders[tx] = p
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:193]: this should be a set operation instead.

File: protocols/mempool/protocol-abstract.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        mempool = append(mempool, tx)
+    else if seen[tx] // Optional improvement
+        senders[tx] = append(senders[tx], q)
+
+upon commit(txs):
+    for all tx in txs
+        seen[tx] = true
+        mempool = remove(mempool, tx) // Optional improvement
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:201]: This is not happening in the reactor.

Comment (otrack) [inline protocols/mempool/protocol-abstract.md:201]: This is not happening in the reactor.

File: protocols/mempool/protocol-abstract.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        mempool = append(mempool, tx)
+
+upon receive <TX, tx> from p:
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        senders[tx] = p
+        mempool = append(mempool, tx)
+    else if seen[tx] // Optional improvement
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:195]: This should not be optional.

File: protocols/mempool/protocol-abstract.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        seen[tx] = true
+        mempool = remove(mempool, tx) // Optional improvement
+
+func send_routine(p):
+    while p in peers
+        tx = mempool.next() // blocks while there is no next
+        if p not in senders[tx]
+            send(<TX, tx>) to p
+
+upon add_peer(p):
```
Comment (otrack) [inline protocols/mempool/protocol-abstract.md:209]: You may also add `remove_peer`.

**Review by otrack — COMMENTED** [review]
Nice work toward understanding the algorithmic structure of the mempool reactor :)

File: protocols/mempool/protocol-abstract.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+var seen  // map of transactions already seen
+
+upon brodcast(tx):
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        for all p in peers
+            send(<TX, tx>) to p
+
+upon receive <TX, tx> from p:
+    if not seen[tx] and valid(tx)
```
[the remark this answers is not in the document]
Comment (hvanz) [reply protocols/mempool/protocol-abstract.md:62]: In the code, the validity of a transaction is defined by the response of `mempool.CheckTx`. But it's true that here is not defined.

File: protocols/mempool/protocol-abstract.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+```
+var peers // other nodes to which we are directly connected
+var seen  // map of transactions already seen
+
+upon brodcast(tx):
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        for all p in peers
+            send(<TX, tx>) to p
```
[the remark this answers is not in the document]
Comment (hvanz) [reply protocols/mempool/protocol-abstract.md:59]: I guess it's the message type.

File: protocols/mempool/protocol-abstract.md (hunk 7 above)
Comment (cason) [reply protocols/mempool/protocol-abstract.md:201]: This is debatable. The point here is that represents _more_ than a refinement. It is an additional feature, of preventing some sendings, enabled by the use of the mempool.

Comment (cason) [reply protocols/mempool/protocol-abstract.md:201]: What? We do not remove a committed transaction from the mempool?

File: protocols/mempool/protocol-abstract.md (hunk 1 above)
Comment (cason) [reply protocols/mempool/protocol-abstract.md:205]: Is that a problem for the pseudo-code?

File: protocols/mempool/protocol-abstract.md (hunk 9 above)
Comment (cason) [reply protocols/mempool/protocol-abstract.md:209]: Yes, I can, although it doesn't add any information. It removes the peer from `peers`, causing the `send_routine` for that peer to leave.

**Review by cason — COMMENTED** [review]
Adding some suggestions from other reviewers to include in a batch commit.

File: protocols/mempool/protocol-abstract.md (hunk 5 above)
Comment (cason) [reply protocols/mempool/protocol-abstract.md:191]: Can `valid(tx)` capture this?

File: protocols/mempool/protocol-abstract.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
+# Transaction dissemination protocol [WIP]
+
+This document describes in high level the mempool protocol, responsible for
```
[the remark this answers is not in the document]
Comment (cason) [reply protocols/mempool/protocol-abstract.md]: Do you think it does not document anything relevant? If so, I can try again.

File: protocols/mempool/protocol-abstract.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+the set $\Pi$ of nodes and a set of edges $(p, q)$ connecting every $p \in \Pi$
+to all its peers $q \in peers[p]$.
+The network overlay is assumed to be connected, meaning that there is path in
+the network overlay graph between every pair of nodes in $\Pi$.
+Since both $\Pi$ and the $peers$ relation are dynamic, some paths between nodes
+may be disrupted, but it is assumed that they are eventually repaired or
+replaced by the underlying p2p communication layer.
+
+Finally, the links connecting a node to its peers are assumed to be reliable
+within a connection.
```
[the remark this answers is not in the document]
Comment (cason) [reply protocols/mempool/protocol-abstract.md:37]: But I can't find a more precise definition.

**Comment by cason** [comment]
Closed in favor of #15.

**Pull request closed without merging**